crates/bevy_scene2/macros/src/bsn/codegen.rs

```diff
@@ -0,0 +1,371 @@
+                            <#type_path as #bevy_scene::PatchGetTemplate>::patch(
+                                move |value| {
+                                    *value = <#type_path as #bevy_ecs::template::GetTemplate>::Template::#function(#args);
+                                }
+                            )
+                        }
+                    }
+                    BsnEntry::ChildrenSceneList(scene_list) => {
+                        let scenes = scene_list.0.to_tokens(bevy_scene, bevy_ecs, bevy_asset);
+                        quote! {
+                            #bevy_scene::RelatedScenes::<#bevy_ecs::hierarchy::ChildOf, _>::new(#scenes)
+                        }
+                    }
+                    BsnEntry::RelatedSceneList(BsnRelatedSceneList { scene_list, relationship_path }) => {
+                        let scenes = scene_list.0.to_tokens(bevy_scene, bevy_ecs, bevy_asset);
+                        quote! {
+                            #bevy_scene::RelatedScenes::<<#relationship_path as #bevy_ecs::relationship::RelationshipTarget>::Relationship, _>::new(
+                                #scenes
+                            )
+                        }
+                    }
+                    BsnEntry::InheritedScene(inherited_scene) => match inherited_scene {
+                        BsnInheritedScene::Asset(lit_str) => {
+                            quote!{#bevy_scene::InheritSceneAsset::from(#lit_str)}
+                        },
+                        BsnInheritedScene::Fn{ function, args}=> quote!{#bevy_scene::InheritScene(#function(#args))},
+                    }
+                    BsnEntry::Name(ident) => {
+                        let name = ident.to_string();
+                        quote! {
+                            <#bevy_ecs::name::Name as PatchGetTemplate>::patch(
+                                move |value| {
+                                    *value = Name(#name.into());
+                                }
+                            )
+                        }
+                    }
+       
+            });
+        }
+
+        quote! {(#(#entries,)*)}
+    }
+}
+
+macro_rules! field_value_type {
+    () => {
+        BsnValue::Expr(_)
+            | BsnValue::Closure(_)
+            | BsnValue::Ident(_)
+            | BsnValue::Lit(_)
+            | BsnValue::Tuple(_)
+    };
+}
+
+impl BsnType {
+    fn to_patch_tokens(
+        &self,
+        bevy_ecs: &Path,
+        bevy_scene: &Path,
+        assignments: &mut Vec<TokenStream>,
+        is_root_template: bool,
+        field_path: &[Member],
+        is_path_ref: bool,
+    ) {
+        let path = &self.path;
+        if !is_root_template {
+            assignments.push(quote! {#bevy_scene::touch_type::<#path>();});
+        }
+        let maybe_deref = is_path_ref.then(|| quote!{*});
+        let maybe_borrow_mut = (!is_path_ref).then(|| quote!{&mut});
+        if let Some(variant) = &self.enum_variant {
```

Comment:
I am particularly pleased and impressed with the enum support, I know they can be tricky to handle in these sorts of macros. But the enum update approach makes sense to me, and I'm interested in playing with it.